### PR TITLE
Fix #2887: create secret of docker-registry

### LINKED ISSUE #2887: How to create secret  of docker-registry
I want to run the command :
kubectl create secret docker-registry harbor-secret --namespace=default --docker-server=http://harbor.inner.com --docker-username=admin --docker-password=Harbor12345

### PR DESCRIPTION
**Pull request by laxmikantbpandhare**
## Description

Fix #2887 
Added utility method for secret creation.

## Type of change

 - [x] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [ ] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [x] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by laxmikantbpandhare** [comment]
PR for #2887 issue on kubernetes-client.

File: kubernetes-client/pom.xml [hunk 1]
```diff
@@ -259,6 +265,37 @@
...
     </dependency>
+      <dependency>
+          <groupId>io.fabric8</groupId>
+          <artifactId>kubernetes-model-core</artifactId>
+          <version>5.1.1</version>
+      </dependency>
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>kubernetes-model-apps</artifactId>
+      <version>5.1.1</version>
```
Comment (oscerd) [inline kubernetes-client/pom.xml:276]: I think the version should be used here.

**Review by oscerd — CHANGES_REQUESTED** [review]
I think there is something wrong with the dependencies declared in the pom.

File: kubernetes-client/pom.xml (hunk 1 above)
Comment (rohanKanojia) [reply kubernetes-client/pom.xml:276]: why are these changes required?

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java [hunk 2]
```diff
@@ -16,24 +16,13 @@
...
 import io.fabric8.kubernetes.client.internal.readiness.Readiness;
 import java.time.Duration;
 import java.time.Instant;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.Comparator;
-import java.util.LinkedHashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (rohanKanojia) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java:25]: Maybe this is due to your IDE setting, Could you please revert to not using wildcard imports?

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java [hunk 3]
```diff
@@ -371,7 +360,34 @@ public static Duration getAge(HasMetadata kubernetesResource) {
       Class<?> listTypeClass = Thread.currentThread().getContextClassLoader().loadClass(type.getName() + "List");
       return (Class<KubernetesResourceList<T>>) listTypeClass;
     } catch (ClassNotFoundException | ClassCastException e) {
-      return CustomResourceList.class;
+      return (Class<? extends KubernetesResourceList>) CustomResourceList.class;
```
Comment (rohanKanojia) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java:363]: why is this change required?

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java [hunk 4]
```diff
@@ -371,7 +360,34 @@ public static Duration getAge(HasMetadata kubernetesResource) {
...
     }
   }
+
+  /**
+   * Create Secret by using username and password.
+   * @param dockerServer
+   * @param username
+   * @param password
```
Comment (rohanKanojia) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java:371]: Could you please add parameter description to javadoc?

File: kubernetes-client/src/test/java/io/fabric8/kubernetes/client/internal/KubernetesResourceUtilTest.java [hunk 5]
```diff
@@ -220,4 +209,17 @@ void testInferListType() {
     assertEquals(CronJobList.class, KubernetesResourceUtil.inferListType(CronJob.class));
     assertEquals(CustomResourceList.class, KubernetesResourceUtil.inferListType(Good.class));
   }
+
+
+  @Test
+  void testcreateDockerRegistrySecret() throws JsonProcessingException {
+    Secret secret = KubernetesResourceUtil.createDockerRegistrySecret("http://harbor.inner.com", "SecretAdmin", "TestingSecret");
+
+    String header = new String(Base64.getDecoder().decode("eyJhdXRocyI6eyJodHRwOi8vaGFyYm9yLmlubmVyLmNvbSI6eyJwYXNzd29yZCI6IlRlc3RpbmdTZWNyZXQiLCJhdXRoIjoiVTJWamNtVjBRV1J0YVc0NlZHVnpkR2x1WjFObFkzSmxkQT09IiwidXNlcm5hbWUiOiJTZWNyZXRBZG1pbiJ9fX0="));
```
Comment (rohanKanojia) [inline kubernetes-client/src/test/java/io/fabric8/kubernetes/client/internal/KubernetesResourceUtilTest.java:218]: You should be getting this value from secret.getData().get(".dockerconfigjson")

File: kubernetes-client/src/test/java/io/fabric8/kubernetes/client/internal/KubernetesResourceUtilTest.java [hunk 6]
```diff
@@ -220,4 +209,17 @@ void testInferListType() {
...
+
+    JSONObject obj = new JSONObject(header);
+
+    assertEquals("SecretAdmin", obj.getJSONObject("auths").getJSONObject("http://harbor.inner.com").getString("username"));
+    assertEquals("TestingSecret", obj.getJSONObject("auths").getJSONObject("http://harbor.inner.com").getString("password"));
```
Comment (rohanKanojia) [inline kubernetes-client/src/test/java/io/fabric8/kubernetes/client/internal/KubernetesResourceUtilTest.java:223]: Would it be possible to avoid using JSON library and just compare strings instead?

File: kubernetes-client/src/test/java/io/fabric8/kubernetes/client/internal/KubernetesResourceUtilTest.java (hunk 5 above)
Comment (laxmikantbpandhare) [reply kubernetes-client/src/test/java/io/fabric8/kubernetes/client/internal/KubernetesResourceUtilTest.java:218]: Yes, Agree. Modified it.

**Comment by laxmikantbpandhare** [comment]
> I think there is something wrong with the dependencies declared in the pom.

No changes in `pom.xml` as inclusion of `JSONObject` code removed completely.

File: kubernetes-client/src/test/java/io/fabric8/kubernetes/client/internal/KubernetesResourceUtilTest.java (hunk 6 above)
Comment (laxmikantbpandhare) [reply kubernetes-client/src/test/java/io/fabric8/kubernetes/client/internal/KubernetesResourceUtilTest.java]: Modified and compared String directly.

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java (hunk 2 above)
Comment (laxmikantbpandhare) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java]: Modified and turn off wildcard imports from intellij.

File: kubernetes-client/pom.xml (hunk 1 above)
Comment (laxmikantbpandhare) [reply kubernetes-client/pom.xml]: removed `pom.xml` completely.

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java (hunk 3 above)
Comment (laxmikantbpandhare) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java]: That was by mistaken. Modified and rebased it with master.

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java (hunk 4 above)
Comment (laxmikantbpandhare) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java]: Added description for each of the `param` field.

**Review by rohanKanojia — APPROVED** [review]

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java [hunk 7]
```diff
@@ -374,4 +380,31 @@ public static Duration getAge(HasMetadata kubernetesResource) {
...
+   * @return an object of Secret
+   */
+  public static Secret createDockerRegistrySecret(String dockerServer, String username, String password) throws JsonProcessingException {
+    Map<String, Object> dockerConfigMap = new HashMap<>();
+    Map<String, Object> auths = new HashMap<>();
+    Map<String, Object> credentials = new HashMap<>();
+    credentials.put("username", username);
+    credentials.put("password", password);
+    String usernameAndPasswordAuth = username + ":" + password;
+    credentials.put("auth", Base64.getEncoder().encodeToString(usernameAndPasswordAuth.getBytes()));
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java:398]:
```suggestion
    credentials.put("auth", Base64.getEncoder().encodeToString(usernameAndPasswordAuth.getBytes(StandardCharsets.UTF_8)));
```

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java [hunk 8]
```diff
@@ -374,4 +380,31 @@ public static Duration getAge(HasMetadata kubernetesResource) {
...
+    auths.put(dockerServer, credentials);
+    dockerConfigMap.put("auths", auths);
+
+    String dockerConfigAsStr = Serialization.jsonMapper().writeValueAsString(dockerConfigMap);
+
+    return new SecretBuilder()
+      .withNewMetadata().withName("harbor-secret").endMetadata()
+      .withType("kubernetes.io/dockerconfigjson")
+      .addToData(".dockerconfigjson", Base64.getEncoder().encodeToString(dockerConfigAsStr.getBytes()))
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java:407]:
```suggestion
      .addToData(".dockerconfigjson", Base64.getEncoder().encodeToString(dockerConfigAsStr.getBytes(StandardCharsets.UTF_8))))
```

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java (hunk 7 above)
Comment (laxmikantbpandhare) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java:398]: Modified and added StandardCharsets.UTF_8.

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java (hunk 8 above)
Comment (laxmikantbpandhare) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesResourceUtil.java:407]: Modified and added StandardCharsets.UTF_8.

**Review by manusa — APPROVED** [review]
LGTM, thx!
